### DISCUSSION TITLE
Triggersequences with M62/M63

### DISCUSSION THREAD

**Discussion by AndreLukas121** [grblHAL/core#523] [Q&A] [answered]
Hi everyone, 

I am trying to implement a triggersequence similary like a row of M-Code M62/M63. 

The aim of the game is that I want to move to a position, e.g. X100 Y100, and set a digital output to high for 1 stepper pulse length at this position. Then I want to trigger all 100 stepper pulses of the x-axis to high for 1 stepper pulse length, the other 99 stepper pulses I have a low signal. 
 
For this use case I have implemented a new M-Code command, with the M-Code template. The new M-Code command uses the linked list, that is also used for the M62 or M63, but I need to store more data like number of pulses or period of one pulse. Therefore, I extended the struct output_command with 2 bool, 2 uint16_t and 2 uint8_t variables (picture 2). In stepper_driver_interrupt_handler I initialize my triggersequence struct (see in picture 1) and a counter in the stepperSetStepOutputs counts the stepper pulses. If I have enough time, the trigger sequence runs well, but with many arcs and many trigger sequences, the motor has a step loss, probably the stepper_driver_interrupt_handler is no longer triggered at the expected times.  

Picture 1: 
![Triggersequence_struct_in_ISR](https://github.com/grblHAL/core/assets/170618283/4f3e39db-6953-409f-8941-bb138743ba3c)

Picture 2:
![Struct_Trigger](https://github.com/grblHAL/core/assets/170618283/627da547-1dab-4215-807a-dcdfb35eb6e4)

Does anyone have any ideas on how I can improve my code so that the ISR stepper_driver_interrupt_handler take less execution time or does anyone have a different approach to implementing a trigger sequence with motion synchronization?

I hope you can help me. Many thanks in advance. 

Best regards 

Andre

**Comment by terjeio** [comment]
Which driver(s) is your target? One? Many?
What are your timing requirements?

A general comment would be not to modify the core, you can "hook in" your code to be executed on the stepper IRQ, either before of after a step is generated. This would allow you to keep all your code in a separate plugin.
[This plugin](https://github.com/grblHAL/Plugin_odometer/blob/master/odometer.c) is one example of how to do it - `stepperPulseStart()` is executed on each step pulse.

**Reply by AndreLukas121** [reply]
Hi Terjeio, 

thank you for the quick response. 
I use two drives for the X and Y axes and the worst-case timing requirements are a pulse periode about 25µsec.

First of all thanks, I think a plugin is a more useful approach for my trigger sequence. 
Now, I have a few questions about the plugin. 

The stepperPulseStart() is very useful for stepper pulse counting, but I also need to trigger synchronized with the motion. 
Therefore, is it possible to add a "hook in" to the stepper_driver_interrupt_handler? Because the ISR is the only place I've seen to execute the output_commands with the linked list. 
Do you know of any other way to get access to the linked list so that the trigger is executed in sync with the motion?  

Thank you in advance, I hope you can help me. 

Best regards

Andre

**Reply by terjeio** [reply]
> Therefore, is it possible to add a "hook in" to the stepper_driver_interrupt_handler?

It is already there? Output commands are executed [here](https://github.com/grblHAL/core/blob/21329cd03fccef58d1c14a23e7587877fc973e63/stepper.c#L327), calling the code stored in the `hal.port.digital_out` function pointer. If you store the current function pointer in your plugin (at startup) and replace it with a pointer to your function you kind of get access to the list. You can then test for the port number(s) you are interested in and set flags or whatever you need to take action on before you pass on the call (if you need to) to the saved function pointer.

**Reply by AndreLukas121** [reply]
Thank you for fast reply.  

If I understand correctly, this is just a list of ports and their values. This is a very good hint that I can get access to it, but I think it is not quite what I need.  Because the question for me is: How do I find out at which position my port should be active and which M code caused this? Because from the G-code and the associated M-code, I would like to specify a position when a trigger/port should be active. At the same time, I use the M code to specify the pulse period and the number of pulses I have to execute. I thought for this purpose I should have access to st.exec_block->output_commands to enable triggering synchronized with the movement. 
For this I must have access to the executed block, or not? 

Best regards

Andre

**Reply by terjeio** [reply, marked as answer]
If you combine your new M-code with M62/M63 I believe you can keep your custom linked list in the plugin. When hal.port.digital_out() gets called set a flag in your plugin and act on that on the next hal.stepper.pulse_start() call.
At the time of your new M-code is parsed get the position from gc_state, save it in your list and later compare that to the position in the hal.stepper.pulse_start() struct parameter when the associated M62/M63 is executed.
If you want you can also add a number of virtual digital outputs and use the port numbers of these as an id if that makes it easier to keep track.

**Reply by AndreLukas121** [reply]
Yes, I think it could work like that. Thanks for the quick help!
